### DISCUSSION TITLE
Handle BulkInsert as Queue

### DISCUSSION THREAD

**Discussion by bobskigit** [ravendb/ravendb#13202] [Q&A]
We potentially have to handle many many thousands of inserts over a short time period in a large web-based app. Is there any way to add the inserts/updates to a queue and place them on a background thread to bulk insert them later?

We are measured on the response time from the users device and the slowest call is the insert/update, if we queue this it will reduce the perceived response time.

**Comment by ayende** [comment]
Bulk insert is optimized for inserting / updating a LOT of values, as soon as possible. Note that it is _not_ safe to use a single bulk insert concurrently.
However, you can put that in front of a queue. The issue is that you'll likely run into problems because you don't have a way to confirm that the operation was completed.
Bulk insert will only guarantee writes after you call `Dispose`, after all.

I don't think that this is a good scenario for you needs. However, this is likely better:
https://ayende.com/blog/176257/10x-speedup-utilizing-nagle-algorithm-in-business-application

The idea is to reduce overall costs in network transfer / disk I/O. You have a small queue that you'll wait to fill until a certain size / time reached. Then you send it to the server (basically, Session + SaveChanges()). You'll report success to everything in the batch when you are done with the batch.
